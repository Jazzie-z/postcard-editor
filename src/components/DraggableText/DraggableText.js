import Draggable from 'react-draggable';
import styled from 'styled-components';
const TextContainer = styled.div`
    position: absolute;
    top: 15%;
    left: 32%;
`
const EditableText = styled.div`
    color: ${({theme})=>theme.colors.white};
    width: fit-content;    
    font-size: ${({ zoom }) => (16 + zoom) > 1 ? (16 + zoom) : 1}px;
    transform: rotate(${({ rotate }) => rotate}deg);
    &:focus{
        background: ${({theme})=>theme.colors.white};
        color: ${({theme})=>theme.colors.black};
        border: none;
        outline:none;
    }
`
const DraggableText = ({ index, title, zoom, rotate, setActiveItem }) => {
    return (
        <Draggable bounds="parent">
            <TextContainer>
                <EditableText
                    contentEditable
                    suppressContentEditableWarning 
                    zoom={zoom}
                    rotate={rotate}
                    onClick={() => setActiveItem(index)}>{title}</EditableText>
            </TextContainer>
        </Draggable>
    )
}

export default DraggableText
