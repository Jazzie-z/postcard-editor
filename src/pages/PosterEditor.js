import React, { useState } from 'react'
import PosterViewer from 'components/PosterViewer/PosterViewer'
import styled from 'styled-components'
import ButtonList from '../components/ButtonList/ButtonList'
import { staticText } from '../constants/resource'
const Container = styled.div`
    padding: 20px;
    display: flex;
`
const rotationValue = 90;
const maxRotationValue = 360;
const zoomValue = 1;
const PosterEditor = () => {
    const [texts, setTexts] = useState([]);
    const [activeItem, setActiveItem] = useState(null);
    const addText = () => {
        if(!texts.length) setActiveItem(0)
        setTexts(prev => [...prev, { title: staticText.newText, zoom: 0, rotate: 0 }])
    }
    const changeZoom = (value) => {
        let payload = [...texts]
        if (payload[activeItem]) {
            payload[activeItem].zoom += value
            setTexts(payload)
        }
    }
    const rotate = () => {
        let payload = [...texts]
        if (payload[activeItem]) {
            payload[activeItem].rotate += rotationValue;
            if (payload[activeItem].rotate >= maxRotationValue) payload[activeItem].rotate = 0
            setTexts(payload)
        }
    }
    return (
        <Container>
            <PosterViewer texts={texts} setActiveItem={setActiveItem} />
            <ButtonList addText={addText} zoomIn={() => changeZoom(zoomValue)} zoomOut={() => changeZoom(-zoomValue)} rotate={rotate} />
        </Container>
    )
}

export default PosterEditor
