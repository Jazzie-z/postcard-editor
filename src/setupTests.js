import { configure, shallow, mount, render } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { ThemeProvider } from 'styled-components';
import { theme } from './theme';

configure({ adapter: new Adapter() });

const wrapChildWithThemeProvider = children => (
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const shallowWithTheme = children =>
    shallow(wrapChildWithThemeProvider(children));
const mountWithTheme = children =>
    mount(wrapChildWithThemeProvider(children));
const renderWithTheme = children =>
    render(wrapChildWithThemeProvider(children));

export { shallowWithTheme, mountWithTheme, renderWithTheme };