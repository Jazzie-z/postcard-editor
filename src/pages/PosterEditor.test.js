
import DraggableText from "components/DraggableText/DraggableText";
import { mountWithTheme } from "../setupTests";
import PosterEditor from "./PosterEditor";

describe('PosterEditor', () => {
    let wrapper;
    beforeEach(()=>{
        wrapper = mountWithTheme(<PosterEditor />)
    })
    afterEach(()=>{
        wrapper.unmount()
    })
    it('renders properly', () => {
        expect(wrapper).toMatchSnapshot()
    })
    it('adds new text block', () => {
        wrapper = mountWithTheme(<PosterEditor />);
        expect(wrapper.find('div').at(1).text()).toBe('')
        wrapper.find('div').at(5).simulate('click')
        expect(wrapper.find(DraggableText).text()).toBe('New Text...')
    })
    it('rotates the text block', () => {
        wrapper = mountWithTheme(<PosterEditor />);
        expect(wrapper.find('div').at(1).text()).toBe('')
        wrapper.find('div').at(5).simulate('click')
        wrapper.find(DraggableText).find('div').at(1).simulate('click')
        wrapper.find('div').at(10).simulate('click')
        expect(wrapper.find(DraggableText).props().rotate).toBe(90)
    })
    it('zooms in the text block', () => {
        wrapper = mountWithTheme(<PosterEditor />);
        expect(wrapper.find('div').at(1).text()).toBe('')
        wrapper.find('div').at(5).simulate('click')
        wrapper.find(DraggableText).find('div').at(1).simulate('click')
        wrapper.find('div').at(8).simulate('click')
        expect(wrapper.find(DraggableText).props().zoom).toBe(1)
    })
})