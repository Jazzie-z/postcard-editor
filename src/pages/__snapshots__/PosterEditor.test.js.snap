// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PosterEditor renders properly 1`] = `ReactWrapper {}`;
